File: Leetcode/1372/1372.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var (
	father map[*TreeNode]*TreeNode
	dp map[*TreeNode]int
)
func dfs(node *TreeNode, fa *TreeNode) {
	if node != nil {
		father[node] = fa
		if node.Left != nil {
			dfs(node.Left, node)
		}
		if node.Right != nil {
			dfs(node.Right, node)
		}
	}
}
func solve(node *TreeNode) int {
	val, ok := dp[node]
	if ok {
		return val
	}
	if father[node] == nil {
		dp[node] = 0
	} else if node == father[node].Left {
		if father[father[node]] == nil {
			dp[node] = 1
		} else if father[node] == father[father[node]].Left {
			dp[node] = 1
		} else {
			dp[node] = dp[father[node]] + 1
		}
	} else {
		if father[father[node]] == nil {
			dp[node] = 1
		} else if father[node] == father[father[node]].Right {
			dp[node] = 1
		} else {
			dp[node] = dp[father[node]] + 1
		}
	}
	return dp[node]
}
func dfs2(node *TreeNode, ans *int) {
	if node != nil {
		temp := solve(node)
		if temp > *ans {
			*ans = temp
		}
		dfs2(node.Left, ans)
		dfs2(node.Right, ans)
	}
}
func longestZigZag(root *TreeNode) int {
	father = nil
	father = make(map[*TreeNode]*TreeNode)
	dfs(root, nil)
	dp = nil
	dp = make(map[*TreeNode]int)
	ans := 0
	dfs2(root, &ans)
	return ans
}

File: Leetcode/1372/1372.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*, TreeNode*> father;
    unordered_map<TreeNode*,int> dp;
    int ans;
    int solve(TreeNode* node) {
        if (dp.find(node) != dp.end()) {
            return dp[node];
        }
        // Case 1: node is root
        if (!father[node]) {
            dp[node] = 0;
        }
        // Case 2: node is the left child
        else if (node == father[node]->left) {
            // Subcase 1: father is root
            if (!father[father[node]]) {
                dp[node] = 1;
            }
            // Subcase 2: father is the right child of its father
            else if (father[node] == father[father[node]]->right) {
                dp[node] = dp[father[node]] + 1;
            }
            // Subcase 3: father is the left child of its father
            else {
                dp[node] = 1;
            }
        }
        // Case 3: node is the right child
        else {
            if (!father[father[node]]) {
                dp[node] = 1;
            }
            else if (father[node] == father[father[node]]->left) {
                dp[node] = dp[father[node]] + 1;
            }
            else {
                dp[node] = 1;
            }
        }
        return dp[node];
    }
    // get the father of each node
    void dfs1(TreeNode* node) {
        if (node->left && node->right) {
            father[node->left] = node;
            father[node->right] = node;
            dfs1(node->left);
            dfs1(node->right);
        }
        else if (node->left && !node->right) {
            father[node->left] = node;
            dfs1(node->left);
        }
        else if (!node->left && node->right) {
            father[node->right] = node;
            dfs1(node->right);
        }
    }
    // get the longest zigzag path
    void dfs2(TreeNode* node) {
        if (node) {
            ans = max(ans, solve(node));
            dfs2(node->left);
            dfs2(node->right);
        }
    }
    int longestZigZag(TreeNode* root) {
        ans = 0;
        dp.clear();
        father.clear();
        father[root] = nullptr;
        dfs1(root);
        dfs2(root);
        return ans;
    }
};